opendatahub-io/data-science-pipelines-operator
# Have pre-req retry.

## The issue resolved by this Pull Request:
Resolves  https://issues.redhat.com/browse/RHOAIENG-2099

## Description of your changes:

Fixes a bug where we were not returning an error where expected, when DB connection failed to query against the configured DB.

Also added a fall back reconcile requeue. When a health check by DSPO fails for Object store or Database, dspo will requeue this DSPA to reconcile again after 20s (by default, but configurable at dspo level), then try again.

## Other considerations

This adds infinite retry logic when prereqs fail. There is no sort of exponential backoff added here, the logic is pretty. My thinking is to iterate on this and eventually add a max retry attempts or some sort of backoff time in the status field.

## Testing instructions

Deploy DSPO
Deploy Multiple DSPA's
Inspect the DSPA status field, ensure the proper status message/reasons are being trickled into the status field
Try with a working DB, then an invalid endpoint.
Try to scale down the default MariaDB before it comes up, and inspect the behavior of the DSPO, then try scaling it back up again. 

## Checklist
